### ISSUE TITLE
Test suite fails zarmour self test (Assertion `zchunk_size (chunk) == strlen (expected) + 1' failed)

### ISSUE THREAD

**Issue by Apteryks** [zeromq/czmq#2125]
Hello,

Attempting to build version 4.2.0 on Guix, the test suite fails with:

```
make[2]: Circular src/selftest-ro <- src/selftest-ro dependency dropped.
/gnu/store/pwcp239kjf7lnj5i4lkdzcfcxwcfyk72-bash-minimal-5.0.16/bin/bash ./libtool --mode=execute ./src/czmq_selftest
czmq_selftest: src/zarmour.c:658: s_armour_decode: Assertion `zchunk_size (chunk) == strlen (expected) + 1' failed.
make[2]: *** [Makefile:2295: check-local] Aborted
make[2]: Leaving directory '/tmp/guix-build-czmq-4.2.0.drv-0/czmq-4.2.0'
make[1]: *** [Makefile:2085: check-am] Error 2
make[1]: Leaving directory '/tmp/guix-build-czmq-4.2.0.drv-0/czmq-4.2.0'
make: *** [Makefile:1641: check-recursive] Error 1
```

Running `check-verbose` in the failed build directory yields:
```
make check-verbose
make: Circular src/selftest-ro <- src/selftest-ro dependency dropped.
/gnu/store/pwcp239kjf7lnj5i4lkdzcfcxwcfyk72-bash-minimal-5.0.16/bin/bash ./libtool --mode=execute ./src/czmq_selftest -v
Running czmq selftests...
 * zactor: OK
 * zarmour: 
D: 20-10-20 23:35:19 zarmour:
D: 20-10-20 23:35:19     mode:        base64
D: 20-10-20 23:35:19     pad:         false
D: 20-10-20 23:35:19     pad_char:    '='
D: 20-10-20 23:35:19     line_breaks: true
D: 20-10-20 23:35:19     line_length: 64
D: 20-10-20 23:35:19     encoded '' into '' ('')
D: 20-10-20 23:35:19     decoded '' into ''
D: 20-10-20 23:35:19     encoded 'f' into 'Zg' ('Zg')
D: 20-10-20 23:35:19     decoded 'Zg' into 'f'
D: 20-10-20 23:35:19     encoded 'fo' into 'Zm8' ('Zm8')
D: 20-10-20 23:35:19     decoded 'Zm8' into 'fo'
D: 20-10-20 23:35:19     encoded 'foo' into 'Zm9v' ('Zm9v')
D: 20-10-20 23:35:19     decoded 'Zm9v' into 'foo'
D: 20-10-20 23:35:19     encoded 'foob' into 'Zm9vYg' ('Zm9vYg')
D: 20-10-20 23:35:19     decoded 'Zm9vYg' into 'foob'
D: 20-10-20 23:35:19     encoded 'fooba' into 'Zm9vYmE' ('Zm9vYmE')
D: 20-10-20 23:35:19     decoded 'Zm9vYmE' into 'fooba'
D: 20-10-20 23:35:19     encoded 'foobar' into 'Zm9vYmFy' ('Zm9vYmFy')
D: 20-10-20 23:35:19     decoded 'Zm9vYmFy' into 'foobar'
D: 20-10-20 23:35:19 zarmour:
D: 20-10-20 23:35:19     mode:        base64
D: 20-10-20 23:35:19     pad:         true
D: 20-10-20 23:35:19     pad_char:    '='
D: 20-10-20 23:35:19     line_breaks: true
D: 20-10-20 23:35:19     line_length: 64
D: 20-10-20 23:35:19     encoded '' into '' ('')
D: 20-10-20 23:35:19     decoded '' into ''
D: 20-10-20 23:35:19     encoded 'f' into 'Zg==' ('Zg==')
D: 20-10-20 23:35:19     decoded 'Zg==' into 'f'
D: 20-10-20 23:35:19     encoded 'fo' into 'Zm8=' ('Zm8=')
D: 20-10-20 23:35:19     decoded 'Zm8=' into 'fo'
D: 20-10-20 23:35:19     encoded 'foo' into 'Zm9v' ('Zm9v')
D: 20-10-20 23:35:19     decoded 'Zm9v' into 'foo'
D: 20-10-20 23:35:19     encoded 'foob' into 'Zm9vYg==' ('Zm9vYg==')
D: 20-10-20 23:35:19     decoded 'Zm9vYg==' into 'foob'
D: 20-10-20 23:35:19     encoded 'fooba' into 'Zm9vYmE=' ('Zm9vYmE=')
D: 20-10-20 23:35:19     decoded 'Zm9vYmE=' into 'fooba'
D: 20-10-20 23:35:19     encoded 'foobar' into 'Zm9vYmFy' ('Zm9vYmFy')
D: 20-10-20 23:35:19     decoded 'Zm9vYmFy' into 'foobar'
D: 20-10-20 23:35:19 zarmour:
D: 20-10-20 23:35:19     mode:        base64url
D: 20-10-20 23:35:19     pad:         false
D: 20-10-20 23:35:19     pad_char:    '='
D: 20-10-20 23:35:19     line_breaks: true
D: 20-10-20 23:35:19     line_length: 64
D: 20-10-20 23:35:19     encoded '' into '' ('')
D: 20-10-20 23:35:19     decoded '' into ''
D: 20-10-20 23:35:19     encoded 'f' into 'Zg' ('Zg')
D: 20-10-20 23:35:19     decoded 'Zg' into 'f'
D: 20-10-20 23:35:19     encoded 'fo' into 'Zm8' ('Zm8')
D: 20-10-20 23:35:19     decoded 'Zm8' into 'fo'
D: 20-10-20 23:35:19     encoded 'foo' into 'Zm9v' ('Zm9v')
D: 20-10-20 23:35:19     decoded 'Zm9v' into 'foo'
D: 20-10-20 23:35:19     encoded 'foob' into 'Zm9vYg' ('Zm9vYg')
D: 20-10-20 23:35:19     decoded 'Zm9vYg' into 'foob'
D: 20-10-20 23:35:19     encoded 'fooba' into 'Zm9vYmE' ('Zm9vYmE')
D: 20-10-20 23:35:19     decoded 'Zm9vYmE' into 'fooba'
D: 20-10-20 23:35:19     encoded 'foobar' into 'Zm9vYmFy' ('Zm9vYmFy')
D: 20-10-20 23:35:19     decoded 'Zm9vYmFy' into 'foobar'
D: 20-10-20 23:35:19 zarmour:
D: 20-10-20 23:35:19     mode:        base64url
D: 20-10-20 23:35:19     pad:         true
D: 20-10-20 23:35:19     pad_char:    '='
D: 20-10-20 23:35:19     line_breaks: true
D: 20-10-20 23:35:19     line_length: 64
D: 20-10-20 23:35:19     encoded '' into '' ('')
D: 20-10-20 23:35:19     decoded '' into ''
D: 20-10-20 23:35:19     encoded 'f' into 'Zg==' ('Zg==')
D: 20-10-20 23:35:19     decoded 'Zg==' into 'f'
D: 20-10-20 23:35:19     encoded 'fo' into 'Zm8=' ('Zm8=')
D: 20-10-20 23:35:19     decoded 'Zm8=' into 'fo'
D: 20-10-20 23:35:19     encoded 'foo' into 'Zm9v' ('Zm9v')
D: 20-10-20 23:35:19     decoded 'Zm9v' into 'foo'
D: 20-10-20 23:35:19     encoded 'foob' into 'Zm9vYg==' ('Zm9vYg==')
D: 20-10-20 23:35:19     decoded 'Zm9vYg==' into 'foob'
D: 20-10-20 23:35:19     encoded 'fooba' into 'Zm9vYmE=' ('Zm9vYmE=')
D: 20-10-20 23:35:19     decoded 'Zm9vYmE=' into 'fooba'
D: 20-10-20 23:35:19     encoded 'foobar' into 'Zm9vYmFy' ('Zm9vYmFy')
D: 20-10-20 23:35:19     decoded 'Zm9vYmFy' into 'foobar'
D: 20-10-20 23:35:19 zarmour:
D: 20-10-20 23:35:19     mode:        base32
D: 20-10-20 23:35:19     pad:         false
D: 20-10-20 23:35:19     pad_char:    '='
D: 20-10-20 23:35:19     line_breaks: true
D: 20-10-20 23:35:19     line_length: 64
D: 20-10-20 23:35:19     encoded '' into '' ('')
D: 20-10-20 23:35:19     decoded '' into ''
D: 20-10-20 23:35:19     encoded 'f' into 'MY' ('MY')
D: 20-10-20 23:35:19     decoded 'MY' into 'f'
D: 20-10-20 23:35:19     encoded 'fo' into 'MZXQ' ('MZXQ')
D: 20-10-20 23:35:19     decoded 'MZXQ' into 'fo'
D: 20-10-20 23:35:19     encoded 'foo' into 'MZXW6' ('MZXW6')
D: 20-10-20 23:35:19     decoded 'MZXW6' into 'foo'
D: 20-10-20 23:35:19     encoded 'foob' into 'MZXW6YQ' ('MZXW6YQ')
D: 20-10-20 23:35:19     decoded 'MZXW6YQ' into 'foob'
D: 20-10-20 23:35:19     encoded 'fooba' into 'MZXW6YTB' ('MZXW6YTB')
D: 20-10-20 23:35:19     decoded 'MZXW6YTB' into 'fooba'
D: 20-10-20 23:35:19     encoded 'foobar' into 'MZXW6YTBOI' ('MZXW6YTBOI')
D: 20-10-20 23:35:19     decoded 'MZXW6YTBOI' into 'foobar'
D: 20-10-20 23:35:19     decoded 'my' into 'f'
D: 20-10-20 23:35:19     decoded 'mzxq' into 'fo'
D: 20-10-20 23:35:19     decoded 'mzxw6' into 'foo'
D: 20-10-20 23:35:19     decoded 'mzxw6yq' into 'foob'
D: 20-10-20 23:35:19     decoded 'mzxw6ytb' into 'fooba'
D: 20-10-20 23:35:19     decoded 'mzxw6ytboi' into 'foobar'
D: 20-10-20 23:35:19 zarmour:
D: 20-10-20 23:35:19     mode:        base32
D: 20-10-20 23:35:19     pad:         true
D: 20-10-20 23:35:19     pad_char:    '='
D: 20-10-20 23:35:19     line_breaks: true
D: 20-10-20 23:35:19     line_length: 64
D: 20-10-20 23:35:19     encoded '' into '' ('')
D: 20-10-20 23:35:19     decoded '' into ''
D: 20-10-20 23:35:19     encoded 'f' into 'MY======' ('MY======')
D: 20-10-20 23:35:19     decoded 'MY======' into 'f'
D: 20-10-20 23:35:19     encoded 'fo' into 'MZXQ====' ('MZXQ====')
D: 20-10-20 23:35:19     decoded 'MZXQ====' into 'fo'
D: 20-10-20 23:35:19     encoded 'foo' into 'MZXW6===' ('MZXW6===')
D: 20-10-20 23:35:19     decoded 'MZXW6===' into 'foo'
D: 20-10-20 23:35:19     encoded 'foob' into 'MZXW6YQ=' ('MZXW6YQ=')
D: 20-10-20 23:35:19     decoded 'MZXW6YQ=' into 'foob'
D: 20-10-20 23:35:19     encoded 'fooba' into 'MZXW6YTB' ('MZXW6YTB')
D: 20-10-20 23:35:19     decoded 'MZXW6YTB' into 'fooba'
D: 20-10-20 23:35:19     encoded 'foobar' into 'MZXW6YTBOI======' ('MZXW6YTBOI======')
D: 20-10-20 23:35:19     decoded 'MZXW6YTBOI======' into 'foobar'
D: 20-10-20 23:35:19     decoded 'my======' into 'f'
D: 20-10-20 23:35:19     decoded 'mzxq====' into 'fo'
D: 20-10-20 23:35:19     decoded 'mzxw6===' into 'foo'
D: 20-10-20 23:35:19     decoded 'mzxw6yq=' into 'foob'
D: 20-10-20 23:35:19     decoded 'mzxw6ytb' into 'fooba'
D: 20-10-20 23:35:19     decoded 'mzxw6ytboi======' into 'foobar'
D: 20-10-20 23:35:19 zarmour:
D: 20-10-20 23:35:19     mode:        base32hex
D: 20-10-20 23:35:19     pad:         false
D: 20-10-20 23:35:19     pad_char:    '='
D: 20-10-20 23:35:19     line_breaks: true
D: 20-10-20 23:35:19     line_length: 64
D: 20-10-20 23:35:19     encoded '' into '' ('')
D: 20-10-20 23:35:19     decoded '' into ''
D: 20-10-20 23:35:19     encoded 'f' into 'CO' ('CO')
D: 20-10-20 23:35:19     decoded 'CO' into 'f'
D: 20-10-20 23:35:19     encoded 'fo' into 'CPNG' ('CPNG')
D: 20-10-20 23:35:19     decoded 'CPNG' into 'fo'
D: 20-10-20 23:35:19     encoded 'foo' into 'CPNMU' ('CPNMU')
D: 20-10-20 23:35:19     decoded 'CPNMU' into 'foo'
D: 20-10-20 23:35:19     encoded 'foob' into 'CPNMUOG' ('CPNMUOG')
D: 20-10-20 23:35:19     decoded 'CPNMUOG' into 'foob'
D: 20-10-20 23:35:19     encoded 'fooba' into 'CPNMUOJ1' ('CPNMUOJ1')
D: 20-10-20 23:35:19     decoded 'CPNMUOJ1' into 'fooba'
D: 20-10-20 23:35:19     encoded 'foobar' into 'CPNMUOJ1E8' ('CPNMUOJ1E8')
D: 20-10-20 23:35:19     decoded 'CPNMUOJ1E8' into 'foobar'
D: 20-10-20 23:35:19     decoded 'co' into 'f'
D: 20-10-20 23:35:19     decoded 'cpng' into 'fo'
D: 20-10-20 23:35:19     decoded 'cpnmu' into 'foo'
D: 20-10-20 23:35:19     decoded 'cpnmuog' into 'foob'
D: 20-10-20 23:35:19     decoded 'cpnmuoj1' into 'fooba'
D: 20-10-20 23:35:19     decoded 'cpnmuoj1e8' into 'foobar'
D: 20-10-20 23:35:19 zarmour:
D: 20-10-20 23:35:19     mode:        base32hex
D: 20-10-20 23:35:19     pad:         true
D: 20-10-20 23:35:19     pad_char:    '='
D: 20-10-20 23:35:19     line_breaks: true
D: 20-10-20 23:35:19     line_length: 64
D: 20-10-20 23:35:19     encoded '' into '' ('')
D: 20-10-20 23:35:19     decoded '' into ''
D: 20-10-20 23:35:19     encoded 'f' into 'CO======' ('CO======')
D: 20-10-20 23:35:19     decoded 'CO======' into 'f'
D: 20-10-20 23:35:19     encoded 'fo' into 'CPNG====' ('CPNG====')
D: 20-10-20 23:35:19     decoded 'CPNG====' into 'fo'
D: 20-10-20 23:35:19     encoded 'foo' into 'CPNMU===' ('CPNMU===')
D: 20-10-20 23:35:19     decoded 'CPNMU===' into 'foo'
D: 20-10-20 23:35:19     encoded 'foob' into 'CPNMUOG=' ('CPNMUOG=')
D: 20-10-20 23:35:19     decoded 'CPNMUOG=' into 'foob'
D: 20-10-20 23:35:19     encoded 'fooba' into 'CPNMUOJ1' ('CPNMUOJ1')
D: 20-10-20 23:35:19     decoded 'CPNMUOJ1' into 'fooba'
D: 20-10-20 23:35:19     encoded 'foobar' into 'CPNMUOJ1E8======' ('CPNMUOJ1E8======')
D: 20-10-20 23:35:19     decoded 'CPNMUOJ1E8======' into 'foobar'
D: 20-10-20 23:35:19     decoded 'co======' into 'f'
D: 20-10-20 23:35:19     decoded 'cpng====' into 'fo'
D: 20-10-20 23:35:19     decoded 'cpnmu===' into 'foo'
D: 20-10-20 23:35:19     decoded 'cpnmuog=' into 'foob'
D: 20-10-20 23:35:19     decoded 'cpnmuoj1' into 'fooba'
D: 20-10-20 23:35:19     decoded 'cpnmuoj1e8======' into 'foobar'
D: 20-10-20 23:35:19 zarmour:
D: 20-10-20 23:35:19     mode:        base16
D: 20-10-20 23:35:19     pad:         true
D: 20-10-20 23:35:19     pad_char:    '='
D: 20-10-20 23:35:19     line_breaks: true
D: 20-10-20 23:35:19     line_length: 64
D: 20-10-20 23:35:19     encoded '' into '' ('')
D: 20-10-20 23:35:19     decoded '' into ''
D: 20-10-20 23:35:19     encoded 'f' into '66' ('66')
D: 20-10-20 23:35:19     decoded '66' into 'f'
D: 20-10-20 23:35:19     encoded 'fo' into '666F' ('666F')
D: 20-10-20 23:35:19     decoded '666F' into 'fo'
D: 20-10-20 23:35:19     encoded 'foo' into '666F6F' ('666F6F')
D: 20-10-20 23:35:19     decoded '666F6F' into 'foo'
D: 20-10-20 23:35:19     encoded 'foob' into '666F6F62' ('666F6F62')
D: 20-10-20 23:35:19     decoded '666F6F62' into 'foob'
D: 20-10-20 23:35:19     encoded 'fooba' into '666F6F6261' ('666F6F6261')
D: 20-10-20 23:35:19     decoded '666F6F6261' into 'fooba'
D: 20-10-20 23:35:19     encoded 'foobar' into '666F6F626172' ('666F6F626172')
D: 20-10-20 23:35:19     decoded '666F6F626172' into 'foobar'
D: 20-10-20 23:35:19     decoded '666f' into 'fo'
D: 20-10-20 23:35:19     decoded '666f6f' into 'foo'
D: 20-10-20 23:35:19     decoded '666f6f62' into 'foob'
D: 20-10-20 23:35:19     decoded '666f6f6261' into 'fooba'
D: 20-10-20 23:35:19     decoded '666f6f626172' into 'foobar'
D: 20-10-20 23:35:19 zarmour:
D: 20-10-20 23:35:19     mode:        z85
D: 20-10-20 23:35:19     pad:         false
D: 20-10-20 23:35:19     pad_char:    '='
D: 20-10-20 23:35:19     line_breaks: false
D: 20-10-20 23:35:19     line_length: 64
D: 20-10-20 23:35:19     encoded '' into '' ('')
D: 20-10-20 23:35:19     decoded '' into 'foobar'
czmq_selftest: src/zarmour.c:658: s_armour_decode: Assertion `zchunk_size (chunk) == strlen (expected) + 1' failed.
make: *** [Makefile:2299: check-verbose] Aborted
```

Thank you!

**Comment by joequant** [comment]
I know why it's happening.  The self-test is taking a "const char *" pointer and writing into it, and this causes problems with some compilers which say "well it's const so it shouldn't change".

The solution would be to have the code write into a writable bytes buffer.

**Comment by joequant** [comment]
This seems to have been fixed by commits

0c7cac3be12707225d03888a6047e5133d926751
ace06a41da51b1196eef411669343cdf7e8665e2

**Comment by Apteryks** [comment]
Hello, I confirm it's fixed in 4.2.1.  Closing.  Thanks!
